<a href="https://colab.research.google.com/github/shail1811/DL/blob/master/AirQuality.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [14]:
import numpy as np
import pandas as pd

In [20]:
dataset = pd.read_excel('/content/AirQualityUCI.xlsx')
dataset = dataset.drop(columns=['Date' ,'Time'] )
dataset

CO(GT)  PT08.S1(CO)  NMHC(GT)  ...       T         RH        AH
0        2.6      1360.00       150  ...  13.600  48.875001  0.757754
1        2.0      1292.25       112  ...  13.300  47.700000  0.725487
2        2.2      1402.00        88  ...  11.900  53.975000  0.750239
3        2.2      1375.50        80  ...  11.000  60.000000  0.786713
4        1.6      1272.25        51  ...  11.150  59.575001  0.788794
...      ...          ...       ...  ...     ...        ...       ...
9352     3.1      1314.25      -200  ...  21.850  29.250000  0.756824
9353     2.4      1162.50      -200  ...  24.325  23.725000  0.711864
9354     2.4      1142.00      -200  ...  26.900  18.350000  0.640649
9355     2.1      1002.50      -200  ...  28.325  13.550000  0.513866
9356     2.2      1070.75      -200  ...  28.500  13.125000  0.502804

[9357 rows x 13 columns]

In [21]:
Humidity = dataset.AH
Humidity

0       0.757754
1       0.725487
2       0.750239
3       0.786713
4       0.788794
          ...   
9352    0.756824
9353    0.711864
9354    0.640649
9355    0.513866
9356    0.502804
Name: AH, Length: 9357, dtype: float64

In [22]:
dataset = dataset.drop(columns=['AH'])
dataset

CO(GT)  PT08.S1(CO)  NMHC(GT)  ...  PT08.S5(O3)       T         RH
0        2.6      1360.00       150  ...      1267.50  13.600  48.875001
1        2.0      1292.25       112  ...       972.25  13.300  47.700000
2        2.2      1402.00        88  ...      1074.00  11.900  53.975000
3        2.2      1375.50        80  ...      1203.25  11.000  60.000000
4        1.6      1272.25        51  ...      1110.00  11.150  59.575001
...      ...          ...       ...  ...          ...     ...        ...
9352     3.1      1314.25      -200  ...      1728.50  21.850  29.250000
9353     2.4      1162.50      -200  ...      1269.00  24.325  23.725000
9354     2.4      1142.00      -200  ...      1092.00  26.900  18.350000
9355     2.1      1002.50      -200  ...       769.75  28.325  13.550000
9356     2.2      1070.75      -200  ...       816.00  28.500  13.125000

[9357 rows x 12 columns]

In [23]:
from sklearn.preprocessing import MinMaxScaler
scale = MinMaxScaler()
dataset = scale.fit_transform(dataset)
dataset

array([[0.95611137, 0.69650631, 0.25197984, ..., 0.53897714, 0.87326247,
        0.86197939],
       [0.95327985, 0.66625739, 0.22462203, ..., 0.43053898, 0.87203598,
        0.85790977],
       [0.95422369, 0.7152584 , 0.20734341, ..., 0.46790928, 0.86631235,
        0.87964326],
       ...,
       [0.95516753, 0.59917401, 0.        , ..., 0.47452025, 0.92763696,
        0.75625595],
       [0.95375177, 0.53689028, 0.        , ..., 0.35616564, 0.93346279,
        0.73963114],
       [0.95422369, 0.56736243, 0.        , ..., 0.37315214, 0.93417825,
        0.73815915]])

In [24]:
Humidity

0       0.757754
1       0.725487
2       0.750239
3       0.786713
4       0.788794
          ...   
9352    0.756824
9353    0.711864
9354    0.640649
9355    0.513866
9356    0.502804
Name: AH, Length: 9357, dtype: float64

In [63]:
from sklearn.decomposition import PCA
pca = PCA(n_components=3)
new_dataset = pca.fit_transform(dataset)
new_dataset

array([[-0.21138365, -0.1833058 , -0.0166851 ],
       [-0.16507   , -0.10169593, -0.11139899],
       [-0.19473858, -0.13523235, -0.07124481],
       ...,
       [-0.27664164, -0.03793361,  0.08641225],
       [-0.23272301,  0.07094719, -0.01257324],
       [-0.25392779,  0.028889  ,  0.02782045]])

In [64]:
pca.get_covariance()

array([[ 1.32704404e-01,  1.66862749e-03,  4.44038564e-03,
        -1.09443198e-03,  5.30058515e-04,  3.03600325e-02,
        -1.96419036e-03,  6.09873637e-02, -5.45852124e-03,
         3.55883461e-03, -3.49566615e-03, -2.00420539e-03],
       [ 1.66862749e-03,  2.49707698e-02,  1.39789007e-03,
         1.94212042e-02,  1.83778792e-02,  6.32017407e-03,
         2.28520900e-03,  6.55972694e-03,  1.88538590e-02,
         2.04150919e-02,  1.98400074e-02,  1.92416388e-02],
       [ 4.44038564e-03,  1.39789007e-03,  5.52183098e-03,
         1.24014496e-03,  1.28404356e-03,  1.53441228e-03,
         2.53174614e-05,  2.60250425e-03,  1.08619982e-03,
         1.54651654e-03,  1.16762198e-03,  1.17956199e-03],
       [-1.09443198e-03,  1.94212042e-02,  1.24014496e-03,
         2.82074667e-02,  1.69482116e-02, -4.28447108e-04,
         7.53236327e-03, -1.82596710e-03,  1.97617580e-02,
         1.71761874e-02,  2.50025381e-02,  2.42109444e-02],
       [ 5.30058515e-04,  1.83778792e-02,  1.2840435

In [65]:
explained_variance=pca.explained_variance_ratio_
explained_variance

array([0.4263395 , 0.35139688, 0.10857925])

In [66]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(new_dataset,Humidity,test_size=0.3)

In [67]:
from keras.models import Sequential
from keras.layers import Dense
from keras.optimizers import SGD,Adam

model = Sequential()

model.add(Dense(300,input_shape=(3,),activation='relu'))

model.add(Dense(200,activation='relu'))
model.add(Dense(100,activation='relu'))
model.add(Dense(50,activation='relu'))
model.add(Dense(10,activation='relu'))
model.add(Dense(1,activation='linear'))

model.compile(Adam(lr=0.01),'mean_squared_error')

model.summary()


Model: "sequential_8"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_47 (Dense)             (None, 300)               1200      
_________________________________________________________________
dense_48 (Dense)             (None, 200)               60200     
_________________________________________________________________
dense_49 (Dense)             (None, 100)               20100     
_________________________________________________________________
dense_50 (Dense)             (None, 50)                5050      
_________________________________________________________________
dense_51 (Dense)             (None, 10)                510       
_________________________________________________________________
dense_52 (Dense)             (None, 1)                 11        
Total params: 87,071
Trainable params: 87,071
Non-trainable params: 0
__________________________________________________

In [68]:
model.fit(X_train,Y_train,epochs=100)

Epoch 1/100
205/205 [==============================] - 0s 2ms/step - loss: 189.6000
Epoch 2/100
205/205 [==============================] - 0s 2ms/step - loss: 0.5898
Epoch 3/100
205/205 [==============================] - 0s 2ms/step - loss: 1.5865
Epoch 4/100
205/205 [==============================] - 0s 2ms/step - loss: 0.5998
Epoch 5/100
205/205 [==============================] - 0s 2ms/step - loss: 278.1802
Epoch 6/100
205/205 [==============================] - 0s 2ms/step - loss: 3.7599
Epoch 7/100
205/205 [==============================] - 0s 2ms/step - loss: 0.0910
Epoch 8/100
205/205 [==============================] - 0s 2ms/step - loss: 0.0872
Epoch 9/100
205/205 [==============================] - 0s 2ms/step - loss: 0.0818
Epoch 10/100
205/205 [==============================] - 0s 2ms/step - loss: 0.0917
Epoch 11/100
205/205 [==============================] - 0s 2ms/step - loss: 0.0826
Epoch 12/100
205/205 [==============================] - 0s 2ms/step - loss: 0.0809
Epoch 13/

In [69]:
result = model.evaluate(X_test,Y_test)
accuracy = 1 - result
accuracy

88/88 [==============================] - 0s 1ms/step - loss: 0.0744


0.9256371334195137

When ,

    Components = 12(All)    

    Accuracy = 0.9734952803701162

    Components = 9

    Accuracy = 0.9838642477989197

    Components = 6

    Accuracy = 0.9638857655227184

    Components = 3

    Accuracy = 0.9256371334195137

